### ISSUE TITLE
deduplication ineffective with my /home

### ISSUE THREAD

**Issue by martin21** [borgbackup/borg#893]
Deduplication is ineffective with my /home. It contains about 150 GiB of mails mostly in maildir, images, tar.gz, git + bzr repos + one larger VirtualBox image and two 1 GiB Amiga emulator harddisk images. Is it about 2,5 million files in total.

I ran initial full backup yesterday and today did incremental and it reported:

 Number of files: 2593114
 
                            Original size      Compressed size    Deduplicated size
     This archive:              138.78 GB            100.97 GB             66.75 GB
     All archives:              443.83 GB            383.65 GB            342.68 GB

Thats almost half of the backed data as deduplicated size and that does seem to be too much for me. Please not tough initially on this run it also backed up encrypted ecryptfs files, but even those have only seen changes from not even one day.

Now it is running again – without ecryptfs files, I backup them seperately for now –, just directly after the first incremental run, and it already reports:

    81.72 GB O 51.13 GB C 31.83 GB D 2227026 N mnt/home-zeit/h...3CANON/IMG_3335.JPG

Thats 32 GiB of new data for 82 GiB of backed up data and IMHO definitely too much.

I see the deduplicated size going up even for larger files which are just unchanged. And that doesn´t seem right for me. Also it does up when digging through the large maildirs with more than a million of mails. I expect it to go up a bit to store the metadata, i.e. which files refer to what data, but about 65 GiB is almost half the size of the actual backed up data.

Thing in my backup script is that I first create a BTRFS snapshot and run borgbackup on that one, so the complete path name to the filename changes everytime, since I add date and time to the snapshot name so its unique. But as far as I understand borgbackup should handle filename changes just fine.

For other filesystems I have proper deduplication results:

    Number of files: 63484

                           Original size      Compressed size    Deduplicated size
    This archive:              247.37 GB            242.06 GB              9.37 MB
    All archives:              691.19 GB            625.71 GB            342.69 GB

(but this one has also way less, larger files)

So it appears to me that at least with my data set in /home the deduplication does not work well, but I do not understand why that is. Also the time to take incremental backups is very long, first incremental took about almost two and a half hours.

2016-04-12T09:13:34+02:00 Sichere /home.
2016-04-12T11:37:59+02:00 /home gesichert.

Anything I can do to help diagnosing why deduplication seems to be so ineffective with my /home? I do expect some changes for Akonadi mysql db + Baloo desktop search, yet… not this much.

I use LZ4 compression in borgbackup and the files are stored on BTRFS Dual SSD RAID 1 with LZO compression in BTRFS, but that compression is transparent so borgbackup sees unencrypted data.

**Comment by enkore** [comment]
* OS
* Borg version

**Comment by martin21** [comment]
During the backup I also see burst writes from borgbackup to the local SSDs at rates of 300 MiB to 1 GiB every then 10 seconds for maybe 20-30 seconds. I bet that these are writes to ~/.cache/borg which is currently 2,5 GiB on this machine.

**Comment by martin21** [comment]
I only see checkpoints in borg list. From the output of my backup script I thought it completed the backup successfully. Anyway I will run it again and check that it completes its work.

Still it should recognize what it already stored and deduplicate that.

I run borgbackup 1.0.0-3~bpo8+1 on backup VM and borgbackup 1.0.1-1 on laptop that I backup.

**Comment by enkore** [comment]
> Still it should recognize what it already stored and deduplicate that.

Yes

> During the backup I also see burst writes from borgbackup to the local SSDs at rates of 300 MiB to 1 GiB every then 10 seconds for maybe 20-30 seconds. I bet that these are writes to ~/.cache/borg which is currently 2,5 GiB on this machine.

Checkpoints are normally created every 300 seconds / 5 minutes. They commit everything, including the cache.

---

That this happens only on one file system, but works for anther sounds strange. There is no corruption or something like that there (check with e.g. sha256sum)?

**Comment by martin21** [comment]
According to scrubbing and recent rsync backup output its all fine:

merkaba:~> btrfs scrub status /home
scrub status for […]
        scrub started at Tue Apr 12 15:40:04 2016 and finished after 00:10:47
        total bytes scrubbed: 297.89GiB with 0 errors

(note that is double the size due to BTRFS RAID 1)

**Comment by martin21** [comment]
The BTRFS is mounted with noatime, maybe that is an issue? But well also that other data BTRFS which works well (but is no BTRFS RAID 1, but single) is also mounted noatime.

**Comment by martin21** [comment]
Borg version: I run borgbackup 1.0.0-3~bpo8+1 on backup VM and borgbackup 1.0.1-1 on laptop that I backup.

OS: Laptop = Backup-Client has Debian Sid/Experimental with self-compiled kernel 4.6-rc2. Backed up filesystem is BTRFS Dual SSD RAID 1, however the cache files are put into a different BTRFS Dual SSD RAID 1 (the one for the Debian installation, aka the / filesystem). Server VM has Debian Jessie.

**Comment by ThomasWaldmann** [comment]
http://borgbackup.readthedocs.org/en/stable/faq.html#i-am-having-troubles-with-some-network-fuse-special-filesystem-why (just assume that btrfs is a "special filesystem" and do like described there).

**Comment by martin21** [comment]
Thomas, BTRFS is a regular *local* filesystem. And the deduplication works on the other BTRFS as well as within a mounted ecryptfs on the /home BTRFS where I see these deduplication issues when I backup the non ecryptfs parts of it. So or so I just don´t have enough space on the SSDs available to copy the data over to an ext4 filesystem for trying. Am I the first user of borgbackup using it with BTRFS?

What I will do first now is see that is actually really completes the backup instead of just a checkpoint. For the Windows image file I now also verified that it has the same inode number in the snapshot than on the original BTRFS subvolume and that the snapshotted version is identical:

merkaba:/mnt/home-zeit> ls -li home-borgbackup-2016-04-13-08:54:27/martin/.VirtualBox/HardDisks 
insgesamt 18514504
11846176 -rw------- 1 martin martin 19065291264 Mär 25 17:18 Winlala.vdi
merkaba:/mnt/home-zeit> ls -li /home/martin/.VirtualBox/HardDisks                              
insgesamt 18514504
11846176 -rw------- 1 martin martin 19065291264 Mär 25 17:18 Winlala.vdi
merkaba:/mnt/home-zeit> cmp home-borgbackup-2016-04-13-08:54:27/martin/.VirtualBox/HardDisks/Winlala.vdi /home/martin/.VirtualBox/HardDisks/Winlala.vdi 
merkaba:/mnt/home-zeit>

Maybe the thing here is that I didn´t notice the backup was just a partial one, except backup script log file tells different and that some stuff was not yet in the repo.

Additionally here is the exact borg call from the script:

DATE=$( date +%Y-%m-%d-%H:%M:%S )
REPO=somerepo-on-remote-server

log "Sichere /home."
SNAP="/mnt/home-zeit/home-borgbackup-$DATE"
[ -d $SNAP ] && btrfs subvol del "$SNAP"
btrfs subvol snap /home "$SNAP"
borg create -p --stats \
        "$REPO::merkaba-home-$DATE" \
        "$SNAP" \
        --compression lz4 \
        --exclude "*/home*/some-mounted-ecryptfs/*" \
        --exclude "*/home*/.some-unmounted-ecryptfs/*" \
        --exclude "*/home*/*/.cache" \
        --exclude "*/home*/kde/build/*" \
        | stdbuf -oL -eL tee -a "$LOG"
log "/home gesichert."
btrfs subvol delete "$SNAP"

**Comment by martin21** [comment]
Do you have any requirement list or test suite on what the backed up filesystem needs to support? Then I´d check that with BTRFS.

**Comment by martin21** [comment]
The current borgbackup run behaves much better:

38.77 GB O 24.11 GB C 1.23 GB D 468002 N mnt/home-zeit/ho...0389808.R807.merkaba

the windows image file is already through… so hopefully this may just be some sort of cache was not yet fully build or backup was incomplete issue.

**Comment by ThomasWaldmann** [comment]
@martin21 the point is that you are having strange issues and we want to make sure they are not caused by your filesystem, thus the hint to reproduce on ext4. we already had some cases when the fs was causing issues (not btrfs in these cases).

**Comment by martin21** [comment]
Thomas, I understand, yet I simply do not have this resources to test with Ext4 on this laptop without replacing at least one of the SSDs by a larger one. I may do this at some time, but not only for testing borgbackup.

Yet, I tried the backup twice again and I think its about going right now:

Time (start): Wed, 2016-04-13 13:18:56
Time (end):   Wed, 2016-04-13 14:56:03
Command line: /usr/bin/borg create -p --stats somerepo::merkaba-home-2016-04-13-13:18:55 /mnt/home-zeit/home-borgbackup-2016-04-13-13:18:55 --compression lz4 --checkpoint-interval 1200 --exclude someexcludes --exclude */home*/*/.cache --exclude */home*/kde/build/*
Number of files: 2479342

                       Original size      Compressed size    Deduplicated size
This archive:              122.98 GB             86.14 GB             32.31 GB
All archives:              702.01 GB            623.97 GB            320.23 GB

                       Unique chunks         Total chunks
Chunk index:                 2596034              5127413


Thats still 32,31 GiB of data, yet Akonadi has a 4,3 GiB database thats frequently changed, baloo has a database, Akonadi receives mails from LKML and other large volume mailing lists and other desktop application may be saving things to the disk, so… this still appears a bit high, but is probably what to expect from running a Plasma session with lots of apps. And its at least half of the original 67 GiB of deduplicated new data.

Time for the backup is still quite long tough. And I do not yet know what the bottleneck is, cause borgbackup uses about 40-80% of one core on the laptop, SSDs are not anyway next 100% utiliziation, network connection is at 10-14% utilization (yet, there may be a bottleneck between network at training center and company network) and I also see no bottleneck on VM, neither for storage there, nor for CPU (about 8% for borgbackup) nor network. However its about 2,5 million of files, so quite some filesystem metadata to process.

If there is no other idea, except for testing with Ext4, I´d close this for now and continue to gather some longer term statistics before looking again whether the deduplication results is within bounds.

**Comment by ThomasWaldmann** [comment]
well, you could just use a ext4 usb3 disk, rsync your /home to there and mount it from there.
some effort, but doesn't require buying bigger internal ssd.

**Comment by ThomasWaldmann** [comment]
you could also try whether --ignore-inode makes a difference (although I'ld assume btrfs has stable inode numbers).

**Comment by enkore** [comment]
Borg from the master branch (!! development version !!) can show which unique chunk count a file has, that could help narrowing down which files are "problematic".

**Comment by ThomasWaldmann** [comment]
if you put the unique chunks to the leftmost place, you can just do `borg ... | sort -nr`

**Comment by enkore** [comment]
`| sort -nrk2` ;-)

**Comment by ThomasWaldmann** [comment]
Any news?

**Comment by martin21** [comment]
Well, I decided to have some more runs before I invest more time into this, and it seems now the  deduplication is working nicely (having, I think 1.0.2 on each side now, using the jessie-backports version on the server).

This is what I had on the last run on /home:

    Time (start): Thu, 2016-04-28 13:18:40
    Time (end):   Thu, 2016-04-28 14:20:08

    Number of files: 2505490
    
                        Original size      Compressed size    Deduplicated size
    This archive:              123.36 GB             86.40 GB              1.64 GB
    All archives:                1.19 TB              1.07 TB            322.77 GB
    
                        Unique chunks         Total chunks
    Chunk index:                 2664772              7932489

And this on the last run of /daten which goes into the same borg repo:

    Time (start): Thu, 2016-04-28 14:23:02
    Time (end):   Thu, 2016-04-28 15:06:58

    Number of files: 63585
    
                        Original size      Compressed size    Deduplicated size
    This archive:              247.48 GB            242.17 GB             15.89 MB
    All archives:                1.44 TB              1.31 TB            322.79 GB
    
                        Unique chunks         Total chunks
    Chunk index:                 2665984              8083740

So we have 123.36 + 247.48 = 370 GiB as the current archives, some older archives in it as well, and the second archive /daten has lots of uncompressible files (jpg and mp3), so I am quite happy with 322.79 GiB deduplicated size for all archives. Also 1,64 GiB of new data for this run of /home I consider to be pretty good, maybe even a bit low :).


I´d still like the backup to go a bit faster. I didn´t see any clear bottleneck, neither CPU, nor network, nor storage is fully utilized throughput. It might be related to latencies, but I bet network latencies are quite short in the networks I ran borgbackup.

I also saw issues with locking of the remote repo, when running several borgbackup commands in a row from my script once. It told me about a lock on the further commands after backuping /home once. I reran it after deleting the lock in the repo on the server and it didn´t happen again so far.

Anyway, I think the dedup issue is gone and the other issues are different issues, so I am completely okay with closing this one. And before opening new ones I´d like to observe a bit more. I intend to report the locking issue if its not already reported and if it happens again.
